Channel array slices incompatible with direction-restricted params

## Summary

When a proc parameter is a direction-annotated channel array (`[]CHAN BYTE in?`), the codegen generates a Go function signature with `[]<-chan byte`. However, callers passing a `[]chan byte` slice fail to compile because Go slices are not covariant — `[]chan byte` is not assignable to `[]<-chan byte`.

## Reproduction

Transpile `sort_pump.occ` with the course module:

```bash
./occam2go -I kroc/modules/course/libsrc -D TARGET.BITS.PER.WORD=32 -o /tmp/sort_pump.go kroc/modules/course/examples/sort_pump.occ
go vet /tmp/sort_pump.go
```

```
vet: /tmp/sort_pump.go:2899:32: cannot use mess (variable of type []chan byte) as []<-chan byte value in argument to screen_plex
```

## Root cause

The occam declaration:
```occam
PROC screen.plex (VAL INT n.inputs,  []CHAN BYTE in?,  CHAN BYTE out!)
```

Generates:
```go
func screen_plex(n_inputs int, in []<-chan byte, out chan<- byte)
```

The caller has:
```occam
[n.screen.inputs]CHAN BYTE mess:
...
screen.plex (n.screen.inputs, mess, out!)
```

Which generates:
```go
mess := make([]chan byte, n_screen_inputs)
// ...
screen_plex(n_screen_inputs, mess, out)  // ERROR: []chan byte ≠ []<-chan byte
```

In Go, individual channels are covariant (`chan T` assignable to `<-chan T`), but slices of channels are not (`[]chan T` is NOT assignable to `[]<-chan T`).

## Affected files

- `sort_pump.occ`
- `sort_pump_2.occ`
- Any program passing channel arrays to direction-restricted parameters

## Possible fixes

1. **Don't apply direction to channel array params** — generate `[]chan byte` instead of `[]<-chan byte` for `[]CHAN BYTE in?`. Simpler but loses Go's compile-time direction safety for channel arrays.
2. **Generate a conversion loop at call sites** — before calling the proc, create a `[]<-chan byte` wrapper slice from the `[]chan byte`, e.g.:
   ```go
   _tmp := make([]<-chan byte, len(mess))
   for i := range mess { _tmp[i] = mess[i] }
   screen_plex(n_screen_inputs, _tmp, out)
   ```

## Related

Tracking issue: #43

--- a/CLAUDE.md
+++ b/CLAUDE.md
@@ -114,6 +114,8 @@ Six packages, one pipeline:
 | `cs[i] ! 42` | `cs[i] <- 42` |
 | `cs[i] ? x` | `x = <-cs[i]` |
 | `PROC f([]CHAN OF INT cs)` | `func f(cs []chan int)` |
+| `PROC f([]CHAN OF INT cs?)` | `func f(cs []chan int)` (direction dropped for array params) |
+| `PROC f([]CHAN OF INT cs!)` | `func f(cs []chan int)` (direction dropped for array params) |
 | `PROC f(CHAN OF INT c?)` | `func f(c <-chan int)` (input/receive-only) |
 | `PROC f(CHAN OF INT c!)` | `func f(c chan<- int)` (output/send-only) |
 | `f(out!, in?)` (call-site dir) | `f(out, in)` (direction annotations ignored) |

--- a/codegen/codegen.go
+++ b/codegen/codegen.go
@@ -2036,7 +2036,7 @@ func (g *Generator) generateProcParams(params []ast.ProcParam) string {
 	for _, p := range params {
 		var goType string
 		if p.IsChanArray {
-			goType = "[]" + chanDirPrefix(p.ChanDir) + g.occamTypeToGo(p.ChanElemType)
+			goType = "[]chan " + g.occamTypeToGo(p.ChanElemType)
 		} else if p.IsChan {
 			goType = chanDirPrefix(p.ChanDir) + g.occamTypeToGo(p.ChanElemType)
 		} else if p.IsOpenArray {

--- a/codegen/codegen_test.go
+++ b/codegen/codegen_test.go
@@ -620,11 +620,11 @@ func TestChanArrayDirParamGen(t *testing.T) {
 `
 	output := transpile(t, input)
 
-	if !strings.Contains(output, "cs []<-chan int") {
-		t.Errorf("expected '[]<-chan int' for input chan array, got:\n%s", output)
+	if !strings.Contains(output, "cs []chan int") {
+		t.Errorf("expected '[]chan int' for input chan array, got:\n%s", output)
 	}
-	if !strings.Contains(output, "out []chan<- int") {
-		t.Errorf("expected '[]chan<- int' for output chan array, got:\n%s", output)
+	if !strings.Contains(output, "out []chan int") {
+		t.Errorf("expected '[]chan int' for output chan array, got:\n%s", output)
 	}
 }
 

--- a/codegen/e2e_array_test.go
+++ b/codegen/e2e_array_test.go
@@ -306,3 +306,30 @@ func TestE2E_MultiAssignmentValues(t *testing.T) {
 		t.Errorf("expected %q, got %q", expected, output)
 	}
 }
+
+func TestE2E_ChanArrayDirParam(t *testing.T) {
+	// Channel arrays passed to direction-annotated params must compile
+	// (Go slices are not covariant, so direction is dropped for array params)
+	occam := `PROC sender([]CHAN OF INT out!)
+  SEQ i = 0 FOR SIZE out
+    out[i] ! i
+:
+PROC receiver([]CHAN OF INT in?)
+  SEQ i = 0 FOR SIZE in
+    INT v:
+    SEQ
+      in[i] ? v
+      print.int(v)
+:
+SEQ
+  [3]CHAN OF INT cs:
+  PAR
+    sender(cs)
+    receiver(cs)
+`
+	output := transpileCompileRun(t, occam)
+	expected := "0\n1\n2\n"
+	if output != expected {
+		t.Errorf("expected %q, got %q", expected, output)
+	}
+}

